### PR TITLE
Update `compile` command to support creating taproot descriptors

### LINKED ISSUE #204: Update `compile` command to support creating taproot descriptors
Currently `tr` script types are not supported but they are an available option in the `rust-miniscript` crate. The `bdk-cli compile` command should support them.

see:
https://docs.rs/miniscript/12.3.2/miniscript/descriptor/enum.Descriptor.html#method.new_tr
https://github.com/bitcoindevkit/bdk-cli/blob/7a71b14740456834c51aae51be8d335537f42a23/src/handlers.rs#L725

### PR DESCRIPTION
**Pull request by va-an**
### Description
Resolves #204.

### Notes to the reviewers
For creating the tr descriptor, I used the NUMS pubkey proposed in [BIP-341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs).

There is [discussion](https://github.com/rust-bitcoin/rust-bitcoin/issues/1322) about adding NUMS key to `rust-bicoin`, we can use it in the future from there.

Also there is [BIP draft](https://github.com/bitcoin/bips/pull/1746) for new descriptor key expression `unspendable()` for exacly this use case - we will simply use descriptor `tr(unspendable(), TREE)`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk-cli/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

### CODE REVIEW

File: src/handlers.rs [hunk 1]
```diff
@@ -72,6 +72,12 @@ use {
     bdk_wallet::chain::{BlockId, CanonicalizationParams, CheckPoint},
 };
 
+/// Well-known unspendable key used for Taproot descriptors when only script path is intended.
+/// This is a NUMS (Nothing Up My Sleeve) point that ensures the key path cannot be used.
```
Comment (tvpeter) [inline src/handlers.rs]: since you provided detailed comments in the usage block, this documentation seems redundant.

File: src/handlers.rs [hunk 2]
```diff
@@ -714,18 +719,31 @@ pub(crate) fn handle_compile_subcommand(
...
     let descriptor = match script_type.as_str() {
         "sh" => Descriptor::new_sh(legacy_policy),
         "wsh" => Descriptor::new_wsh(segwit_policy),
         "sh-wsh" => Descriptor::new_sh_wsh(segwit_policy),
+        "tr" => {
+            // For tr descriptors, we use a well-known unspendable key (NUMS point).
+            // This ensures the key path is effectively disabled and only script path can be used.
+            // See https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs
+            let tree = TapTree::Leaf(Arc::new(taproot_policy));
+            Descriptor::new_tr(NUMS_UNSPENDABLE_KEY.to_string(), Some(tree))
```
Comment (tvpeter) [inline src/handlers.rs]: I guess the NUMS_UNSPENDABLE_KEY represents XonlyPublicKey, so it is better to parse it into an XonlyPublicKey and not used as a string (since strings are used for key placeholders) to avoid type mismatch.

File: src/handlers.rs [hunk 3]
```diff
@@ -714,18 +719,31 @@ pub(crate) fn handle_compile_subcommand(
...
+        }
         _ => panic!("Invalid type"),
```
Comment (tvpeter) [inline src/handlers.rs]: can you help and change from panic to a graceful handling of the error?

File: src/handlers.rs [hunk 4]
```diff
@@ -995,4 +1012,39 @@ mod test {
...
+        let result = handle_compile_subcommand(
+            Network::Testnet,
+            "and(pk(A),pk(B))".to_string(),
+            "tr".to_string(),
+        );
+        assert!(result.is_ok());
+        let json_result = result.unwrap();
+        let descriptor = json_result.get("descriptor").unwrap().as_str().unwrap();
+        assert_eq!(descriptor, EXPECTED_AND_AB);
+    }
```
Comment (tvpeter) [inline src/handlers.rs:1466]: can you also add test cases for invalid/invalid policies and other edge cases?

File: src/handlers.rs (hunk 1 above)
Comment (vadim-anfv) [reply src/handlers.rs]: Deleted.

File: src/handlers.rs (hunk 2 above)
Comment (vadim-anfv) [reply src/handlers.rs]: I've added validation that `NUMS_UNSPENDABLE_KEY` is parsed as `XOnlyPublicKey`.
Do you mean that nums_key should be passed as `XOnlyPublicKey` to the `new_tr` function?

File: src/handlers.rs (hunk 3 above)
Comment (vadim-anfv) [reply src/handlers.rs]: Changed, pls have a look.

File: src/handlers.rs (hunk 4 above)
Comment (vadim-anfv) [reply src/handlers.rs:1466]: Added some cases as separated test.

File: src/handlers.rs [hunk 5]
```diff
@@ -714,19 +719,41 @@ pub(crate) fn handle_compile_subcommand(
     policy: String,
     script_type: String,
 ) -> Result<serde_json::Value, Error> {
+    use bdk_wallet::miniscript::descriptor::TapTree;
+    use std::sync::Arc;
```
Comment (tvpeter) [inline src/handlers.rs]: Please move the imports to the top of the file and gate them.

File: src/handlers.rs [hunk 6]
```diff
@@ -995,4 +1021,79 @@ mod test {
...
+        use bdk_wallet::bitcoin::Network;
+
+        // Expected taproot descriptors with checksums
+        const EXPECTED_PK_A: &str =
+            "tr(50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0,pk(A))#a2mlskt0";
+        const EXPECTED_AND_AB: &str = "tr(50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0,and_v(v:pk(A),pk(B)))#sfplm6kv";
+
+        // Verify our test expectations use the same NUMS key
+        assert!(EXPECTED_PK_A.contains(NUMS_UNSPENDABLE_KEY_HEX));
+        assert!(EXPECTED_AND_AB.contains(NUMS_UNSPENDABLE_KEY_HEX));
```
Comment (chaitika) [inline src/handlers.rs]: I am wondering how much value these assert statements add when we are in control of the expected values ?

File: src/handlers.rs [hunk 7]
```diff
@@ -1022,12 +1026,31 @@ pub(crate) fn handle_compile_subcommand(
...
+            // For tr descriptors, we use a well-known unspendable key (NUMS point).
+            // This ensures the key path is effectively disabled and only script path can be used.
+            // See https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs
+
+            // Validate that the NUMS key is a valid XOnlyPublicKey
+            XOnlyPublicKey::from_str(NUMS_UNSPENDABLE_KEY_HEX)
+                .map_err(|e| Error::Generic(format!("Invalid NUMS key: {e}")))?;
+
+            let tree = TapTree::Leaf(Arc::new(taproot_policy));
+            Descriptor::new_tr(NUMS_UNSPENDABLE_KEY_HEX.to_string(), Some(tree))
```
Comment (tvpeter) [inline src/handlers.rs]: I thought you will use the parsed value. Something like:

```rust
let xonly_public_key =  XOnlyPublicKey::from_str(NUMS_UNSPENDABLE_KEY_HEX)
                .map_err(|e| Error::Generic(format!("Invalid NUMS key: {e}")))?;
              
let tree = TapTree::Leaf(Arc::new(taproot_policy));
            Descriptor::new_tr(xonly_public_key.to_string(), Some(tree))

File: src/handlers.rs [hunk 8]
```diff
@@ -1355,4 +1378,94 @@ mod test {
...
         assert!(is_final(&full_signed_psbt).is_ok());
     }
+
+    #[cfg(feature = "compiler")]
```
Comment (tvpeter) [inline src/handlers.rs:1380]: These `compiler` tests may not run even when the `compiler` feature is enabled because the whole test mod is behind the backend gate. 

You can move the backend gate to only apply to the `test_psbt_is_final` test i.e.

```rust
#[cfg(any(
        feature = "electrum",
        feature = "esplora",
        feature = "cbf",
        feature = "rpc"
    ))]
#[test]
fn test_psbt_is_final() {}  
```

Put the imports too under their respective gates.

File: src/handlers.rs (hunk 6 above)
Comment (vadim-anfv) [reply src/handlers.rs]: Agreed, fixed!

**Comment by va-an** [comment]
> tACK [04d0301](https://github.com/bitcoindevkit/bdk-cli/commit/04d0301358b0149a1061f677a2749e63ac844937)
> 
> Thank you @va-an for working on this.
> 
> nit: If you can squash your commits into one or two messages, it would be nice.

Right, forgot about rebase. 
Did it, now there's 1 commit in the branch.

**Comment by va-an** [comment]
> Thank you @va-an

Thank you for your patience and help!

File: src/handlers.rs [hunk 9]
```diff
@@ -1022,12 +1026,30 @@ pub(crate) fn handle_compile_subcommand(
...
         "sh" => Descriptor::new_sh(legacy_policy),
         "wsh" => Descriptor::new_wsh(segwit_policy),
         "sh-wsh" => Descriptor::new_sh_wsh(segwit_policy),
-        _ => panic!("Invalid type"),
+        "tr" => {
+            // For tr descriptors, we use a well-known unspendable key (NUMS point).
+            // This ensures the key path is effectively disabled and only script path can be used.
+            // See https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs
+
+            let xonly_public_key = XOnlyPublicKey::from_str(NUMS_UNSPENDABLE_KEY_HEX)
```
Comment (notmandatory) [inline src/handlers.rs:1042]: I believe technically the `NUMS_UNSPENDABLE_KEY_HEX` is not supposed to be used as is (privacy concerns), though for now it's fine. Per [BIP-0431](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs), should add an issue to do the following:

"In order to avoid leaking the information that key path spending is not possible it is recommended to pick a fresh integer r in the range 0...n-1 uniformly at random and use H + rG as internal key. It is possible to prove that this internal key does not have a known discrete logarithm with respect to G by revealing r to a verifier who can then reconstruct how the internal key was created."
